### PR TITLE
Add JMP32 instruction support

### LINKED ISSUE #243: Unsupported BPF_JMP32 opcode causes failure of unmarshal in verifier
The current ebpf-verifier doesn't appear to support `BPF_JMP32_REG`/`BPF_JMP32_IMM`instruction parsing.

When we construct the `BPF_JMP32_REG` instruction, as is defined in `bpf-instruction.h` of Linux kernel:
```
#define BPF_JMP32_REG(OP, DST, SRC, OFF)            \
    ((struct bpf_insn) {                    \
        .code  = BPF_JMP32 | BPF_OP(OP) | BPF_X,    \
        .dst_reg = DST,                    \
        .src_reg = SRC,                    \
        .off   = OFF,                    \
        .imm   = 0 })
```
The opcode should be ` BPF_JMP32 | BPF_OP(OP) | BPF_X`.  

Considering that the `OP` is `BPF_JLT` as `0xa0` which defined in https://github.com/torvalds/linux/blob/50be9417e23af5a8ac860d998e1e3f06b8fd79d7/include/uapi/linux/bpf.h#L38

The `BPF_JMP32` macro is `0x6` which defined in
https://github.com/torvalds/linux/blob/50be9417e23af5a8ac860d998e1e3f06b8fd79d7/include/uapi/linux/bpf.h#L17

And `BPF_X` macro is `0x08` which defined in https://github.com/torvalds/linux/blob/50be9417e23af5a8ac860d998e1e3f06b8fd79d7/include/uapi/linux/bpf_common.h#L51

The `BPF_OP` is defined 
 in https://github.com/torvalds/linux/blob/50be9417e23af5a8ac860d998e1e3f06b8fd79d7/include/uapi/linux/bpf_common.h#L31
```
#define BPF_OP(code)    ((code) & 0xf0)
```
Hence the opcode is
```
0x6 |  (0xa0 & 0xf0) | 0x6 = 0xa6
```

When the instruction opcode firstly parsed in `unmarshal` function, it will fetch this `0xa6 & INST_CLS_MASK` opcode as `0x6` and fall into the `INST_CLS_UNUSED` cases which throws `invalid class 0x6` exception.
https://github.com/vbpf/ebpf-verifier/blob/cc3108a4c6c1c6c3a5ba38dd46cfee1c8fed4560/src/asm_unmarshal.cpp#L373-L402

Hence if we pass the below example `BPF_JMP32_IMM` instruction  bytecode to the verifier, it fails with `unmarshaling error at 19: invalid class 0x6` exception.
```
// if w5 < 0x7fffffff goto pc+1
a6 05 01 00 ff ff ff 7f 13 
```

Wondering whether the lack of support for `BPF_JMP32` is expected in the design of the ebpf-verifier, while the Linux kernel verifier supports it.

Thanks.

### PR DESCRIPTION
**Pull request by dthaler**
JMP32 support makes 22 more eBPF conformance tests pass, as well as some cilium tests

Fixes #243

### CODE REVIEW

File: src/crab/ebpf_domain.cpp [hunk 1]
```diff
@@ -108,16 +111,90 @@ static std::vector<linear_constraint_t> bit_jmp_to_cst_interval(const NumAbsDoma
...
+        int32_t dst_lb_value = (int32_t)dst_lb.value().cast_to_sint64();
+        int32_t dst_ub_value = (int32_t)dst_ub.value().cast_to_sint64();
+        src_int_value = (int32_t)src_int_value;
+        switch (op) {
+        case Op::EQ:
+            if (dst_lb_value == src_int_value && dst_ub_value == src_int_value)
+                return {linear_constraint_t::TRUE()};
+            if (dst_lb_value > src_int_value || dst_ub_value < src_int_value)
+                return {linear_constraint_t::FALSE()};
+            return {};
```
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:145]: Returning an empty set and returning TRUE mean the same thing. I assume the choice here is for the sake of symmetry?

**Comment by dthaler** [comment]
@elazarg any other comments or concerns on this PR?

File: src/ebpf_yaml.hpp [hunk 2]
```diff
@@ -33,6 +33,12 @@ struct Failure {
 void print_failure(const Failure& failure, std::ostream& out);
 
 std::optional<Failure> run_yaml_test_case(const TestCase& test_case);
-std::tuple<bool, std::optional<uint64_t>> run_conformance_test_case(const std::vector<uint8_t>& memory_bytes, const std::vector<uint8_t>& program_bytes, bool debug);
+
+struct ConformanceTestResult {
+    bool success;
```
Comment (elazarg) [inline src/ebpf_yaml.hpp:38]:
```suggestion
    bool success{};
```

File: src/ebpf_yaml.cpp [hunk 3]
```diff
@@ -322,10 +322,16 @@ std::tuple<bool, std::optional<uint64_t>> run_conformance_test_case(const std::v
 
         for (const std::string& invariant : actual_last_invariant.value()) {
             if (invariant.rfind("r0.value=", 0) == 0) {
-                return {result, std::stoull(invariant.substr(9))};
+                int64_t lb, ub;
+                if (invariant[9] == '[') {
+                    lb = std::stoull(invariant.substr(10));
+                    ub = std::stoull(invariant.substr(invariant.find(",", 10) + 1));
+                } else
+                    lb = ub = std::stoull(invariant.substr(9));
```
Comment (elazarg) [inline src/ebpf_yaml.cpp:330]:
```suggestion
                } else {
                    lb = ub = std::stoull(invariant.substr(9));
                }
```

**Review by elazarg — APPROVED** [review]
Looks fine. Only a couple of minor comments.

---

I have an issue with all the special casing of finite width in `ebpf_domain`. There should be a dedicated numerical domain for this, and a dedicated flag in the constraints. I hope to get to it at some point.

**Pull request merged**